hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindTargetGreedy.java

```diff
@@ -44,37 +46,64 @@
 
   private ContainerManager containerManager;
   private PlacementPolicy placementPolicy;
+  private Map<DatanodeDetails, Long> sizeEnteringNode;
+  private NodeManager nodeManager;
+  private ContainerBalancerConfiguration config;
+  private Double upperLimit;
+  private List<DatanodeUsageInfo> potentialTargets;
 
   public FindTargetGreedy(
       ContainerManager containerManager,
-      PlacementPolicy placementPolicy) {
+      PlacementPolicy placementPolicy,
+      NodeManager nodeManager) {
+    sizeEnteringNode = new HashMap<>();
     this.containerManager = containerManager;
     this.placementPolicy = placementPolicy;
+    this.nodeManager = nodeManager;
+  }
+
+  private void setUpperLimit(Double upperLimit){
+    this.upperLimit = upperLimit;
+  }
+
+  private void setPotentialTargets(List<DatanodeUsageInfo> potentialTargets) {
+    this.potentialTargets = potentialTargets;
+    sizeEnteringNode.clear();
+    potentialTargets.forEach(
+        p -> sizeEnteringNode.put(p.getDatanodeDetails(), 0L));
+  }
+
+  private void setConfiguration(ContainerBalancerConfiguration conf) {
+    this.config = conf;
   }
 
   /**
    * Find a {@link ContainerMoveSelection} consisting of a target and
    * container to move for a source datanode. Favours more under-utilized nodes.
    * @param source Datanode to find a target for
-   * @param potentialTargets Collection of potential target datanodes
    * @param candidateContainers Set of candidate containers satisfying
    *                            selection criteria
    *                            {@link ContainerBalancerSelectionCriteria}
-   * @param canSizeEnterTarget A functional interface whose apply
    * (DatanodeDetails, Long) method returns true if the size specified in the
    * second argument can enter the specified DatanodeDetails node
    * @return Found target and container
    */
   @Override
   public ContainerMoveSelection findTargetForContainerMove(
-      DatanodeDetails source, Collection<DatanodeDetails> potentialTargets,
-      Set<ContainerID> candidateContainers,
-      BiFunction<DatanodeDetails, Long, Boolean> canSizeEnterTarget) {
-    for (DatanodeDetails target : potentialTargets) {
+      DatanodeDetails source, Set<ContainerID> candidateContainers) {
+    potentialTargets.sort((a, b) -> {
```

Comment:
Let's do a priority queue implementation here as well.

Reply:
i think we should use TreeSet here.
in `FindSourceGreedy`, we always try to get the top one source, and if no target and be matched with it , we will remove it.
but here, we never remove any target , and just try to find a target with Middle-order traversal。but we can not do  traversal on priorityqueue.